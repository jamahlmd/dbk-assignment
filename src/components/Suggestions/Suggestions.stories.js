import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import StoryContainer from '../StoryContainer';
import Suggestions from './index';

const suggestions = [
    {
        "searchterm": "heren truien",
        "nrResults": 1100
    },
    {
        "searchterm": "dames truien",
        "nrResults": 1501
    },
    {
        "searchterm": "kenzo trui",
        "nrResults": 62
    }
];

storiesOf('Suggestions', module)
    .add('Not showing if no suggestions', () =>
        <StoryContainer><Suggestions query="trui" suggestions={[]} handlePickSuggestion={action('picked suggestion')} /></StoryContainer>)
    .add('Showing suggestions query=trui', () =>
        <StoryContainer><Suggestions query="trui" suggestions={suggestions} handlePickSuggestion={action('picked suggestion')} /></StoryContainer>)
    .add('Showing suggestions query=heren', () =>
        <StoryContainer><Suggestions query="heren" suggestions={suggestions} handlePickSuggestion={action('picked suggestion')} /></StoryContainer>)
    ;