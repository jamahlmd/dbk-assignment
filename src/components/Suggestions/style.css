.suggestions {
    padding: 0;
    margin: 0 10px 10px 10px;
    list-style: none;
    border: 1px solid #e6e6e6;
    border-top: none;
}

.suggestions--li {
    padding: 7px 9px 7px 9px;
    border-top: 1px solid #e6e6e6;
}

.suggestions--li:first-child {
    border-top: none;
}

.suggestions--li--button {
    border: none;
    outline: none;
    border-radius: 0;
    padding: 0;
    letter-spacing: .3px;
    font-weight: 800;
    background-color: transparent;
    width: 100%;
    text-align: left;
}

.suggestions--li--button--span {
    font-weight: 300;
}