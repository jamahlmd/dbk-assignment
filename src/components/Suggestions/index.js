import React from 'react';
import PropTypes from 'prop-types';
import './style.css';

const Suggestions = ({ query, suggestions, handlePickSuggestion }) => {
    if (query.length > 2 && suggestions.length > 0) {
        return (
            <ul className="suggestions">
                {
                    suggestions.map((suggestion, i) => {
                        const indexOfQuery = suggestion.searchterm.search(query);
                        if (indexOfQuery !== -1) {
                            const splittedSearchTerm = suggestion.searchterm.split(query);
                            suggestion.adjustedSearchTerm = <span>
                                {splittedSearchTerm[0]}
                                <span className="suggestions--li--button--span">{query}</span>
                                {splittedSearchTerm[1]}</span>;
                        } else {
                            suggestion.adjustedSearchTerm = suggestion.searchterm;
                        }
                        return (
                            <li
                                key={i}
                                className="suggestions--li"
                            >
                                <button
                                    className="suggestions--li--button"
                                    onClick={() => handlePickSuggestion(suggestion.searchterm)}
                                >
                                    {suggestion.adjustedSearchTerm}{' '}
                                    <span className="suggestions--li--button--span">
                                    ({suggestion.nrResults})</span>
                                </button>
                            </li>
                        );

                    })
                }
            </ul>
        )
    }
    return null;
};

Suggestions.propTypes = {
    query: PropTypes.string.isRequired,
    suggestions: PropTypes.arrayOf(PropTypes.shape({
        searchterm: PropTypes.string.isRequired,
        nrResults: PropTypes.number.isRequired,
    })).isRequired,
    handlePickSuggestion: PropTypes.func.isRequired,
};

export default Suggestions;