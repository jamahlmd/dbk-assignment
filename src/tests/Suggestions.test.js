import React from 'react';
import { shallow } from 'enzyme';
import Suggestions from '../components/Suggestions';

const shortQuery = 'tr';
const longQuery = 'trui';
const suggestions = [
    {
        "searchterm": "heren truien",
        "nrResults": 1100
    },
    {
        "searchterm": "dames truien",
        "nrResults": 1501
    }
];

const handlePickSuggestion = () => {};

it('Render no suggestions when query.length < 2', () => {
    const wrapper = shallow(<Suggestions
        query={shortQuery}
        suggestions={suggestions}
        handlePickSuggestion={handlePickSuggestion}
    />);
    expect(wrapper.find('ul').children().length).toBe(0);
    expect(wrapper).toMatchSnapshot();
});

it('Render suggestions when query.length > 2', () => {
    const wrapper = shallow(<Suggestions
        query={longQuery}
        suggestions={suggestions}
        handlePickSuggestion={handlePickSuggestion}
    />);
    expect(wrapper.find('ul').children().length).toBe(2);
    expect(wrapper).toMatchSnapshot();
});