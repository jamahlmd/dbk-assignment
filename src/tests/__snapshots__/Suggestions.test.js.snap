// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render no suggestions when query.length < 2 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Suggestions
    handlePickSuggestion={[Function]}
    query="tr"
    suggestions={
        Array [
            Object {
              "nrResults": 1100,
              "searchterm": "heren truien",
            },
            Object {
              "nrResults": 1501,
              "searchterm": "dames truien",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Render suggestions when query.length > 2 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Suggestions
    handlePickSuggestion={[Function]}
    query="trui"
    suggestions={
        Array [
            Object {
              "adjustedSearchTerm": <span>
                heren 
                <span
                        className="suggestions--li--button--span"
                >
                        trui
                </span>
                en
        </span>,
              "nrResults": 1100,
              "searchterm": "heren truien",
            },
            Object {
              "adjustedSearchTerm": <span>
                dames 
                <span
                        className="suggestions--li--button--span"
                >
                        trui
                </span>
                en
        </span>,
              "nrResults": 1501,
              "searchterm": "dames truien",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li
          className="suggestions--li"
>
          <button
                    className="suggestions--li--button"
                    onClick={[Function]}
          >
                    <span>
                              heren 
                              <span
                                        className="suggestions--li--button--span"
                              >
                                        trui
                              </span>
                              en
                    </span>
                     
                    <span
                              className="suggestions--li--button--span"
                    >
                              (
                              1100
                              )
                    </span>
          </button>
</li>,
        <li
          className="suggestions--li"
>
          <button
                    className="suggestions--li--button"
                    onClick={[Function]}
          >
                    <span>
                              dames 
                              <span
                                        className="suggestions--li--button--span"
                              >
                                        trui
                              </span>
                              en
                    </span>
                     
                    <span
                              className="suggestions--li--button--span"
                    >
                              (
                              1501
                              )
                    </span>
          </button>
</li>,
      ],
      "className": "suggestions",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="suggestions--li--button"
            onClick={[Function]}
>
            <span>
                        heren 
                        <span
                                    className="suggestions--li--button--span"
                        >
                                    trui
                        </span>
                        en
            </span>
             
            <span
                        className="suggestions--li--button--span"
            >
                        (
                        1100
                        )
            </span>
</button>,
          "className": "suggestions--li",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span>
                heren 
                <span
                                className="suggestions--li--button--span"
                >
                                trui
                </span>
                en
</span>,
              " ",
              <span
                className="suggestions--li--button--span"
>
                (
                1100
                )
</span>,
            ],
            "className": "suggestions--li--button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "heren ",
                  <span
                    className="suggestions--li--button--span"
>
                    trui
</span>,
                  "en",
                ],
              },
              "ref": null,
              "rendered": Array [
                "heren ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "trui",
                    "className": "suggestions--li--button--span",
                  },
                  "ref": null,
                  "rendered": "trui",
                  "type": "span",
                },
                "en",
              ],
              "type": "span",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "(",
                  1100,
                  ")",
                ],
                "className": "suggestions--li--button--span",
              },
              "ref": null,
              "rendered": Array [
                "(",
                1100,
                ")",
              ],
              "type": "span",
            },
          ],
          "type": "button",
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="suggestions--li--button"
            onClick={[Function]}
>
            <span>
                        dames 
                        <span
                                    className="suggestions--li--button--span"
                        >
                                    trui
                        </span>
                        en
            </span>
             
            <span
                        className="suggestions--li--button--span"
            >
                        (
                        1501
                        )
            </span>
</button>,
          "className": "suggestions--li",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span>
                dames 
                <span
                                className="suggestions--li--button--span"
                >
                                trui
                </span>
                en
</span>,
              " ",
              <span
                className="suggestions--li--button--span"
>
                (
                1501
                )
</span>,
            ],
            "className": "suggestions--li--button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "dames ",
                  <span
                    className="suggestions--li--button--span"
>
                    trui
</span>,
                  "en",
                ],
              },
              "ref": null,
              "rendered": Array [
                "dames ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "trui",
                    "className": "suggestions--li--button--span",
                  },
                  "ref": null,
                  "rendered": "trui",
                  "type": "span",
                },
                "en",
              ],
              "type": "span",
            },
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "(",
                  1501,
                  ")",
                ],
                "className": "suggestions--li--button--span",
              },
              "ref": null,
              "rendered": Array [
                "(",
                1501,
                ")",
              ],
              "type": "span",
            },
          ],
          "type": "button",
        },
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li
            className="suggestions--li"
>
            <button
                        className="suggestions--li--button"
                        onClick={[Function]}
            >
                        <span>
                                    heren 
                                    <span
                                                className="suggestions--li--button--span"
                                    >
                                                trui
                                    </span>
                                    en
                        </span>
                         
                        <span
                                    className="suggestions--li--button--span"
                        >
                                    (
                                    1100
                                    )
                        </span>
            </button>
</li>,
          <li
            className="suggestions--li"
>
            <button
                        className="suggestions--li--button"
                        onClick={[Function]}
            >
                        <span>
                                    dames 
                                    <span
                                                className="suggestions--li--button--span"
                                    >
                                                trui
                                    </span>
                                    en
                        </span>
                         
                        <span
                                    className="suggestions--li--button--span"
                        >
                                    (
                                    1501
                                    )
                        </span>
            </button>
</li>,
        ],
        "className": "suggestions",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="suggestions--li--button"
              onClick={[Function]}
>
              <span>
                            heren 
                            <span
                                          className="suggestions--li--button--span"
                            >
                                          trui
                            </span>
                            en
              </span>
               
              <span
                            className="suggestions--li--button--span"
              >
                            (
                            1100
                            )
              </span>
</button>,
            "className": "suggestions--li",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span>
                  heren 
                  <span
                                    className="suggestions--li--button--span"
                  >
                                    trui
                  </span>
                  en
</span>,
                " ",
                <span
                  className="suggestions--li--button--span"
>
                  (
                  1100
                  )
</span>,
              ],
              "className": "suggestions--li--button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "heren ",
                    <span
                      className="suggestions--li--button--span"
>
                      trui
</span>,
                    "en",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "heren ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "trui",
                      "className": "suggestions--li--button--span",
                    },
                    "ref": null,
                    "rendered": "trui",
                    "type": "span",
                  },
                  "en",
                ],
                "type": "span",
              },
              " ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "(",
                    1100,
                    ")",
                  ],
                  "className": "suggestions--li--button--span",
                },
                "ref": null,
                "rendered": Array [
                  "(",
                  1100,
                  ")",
                ],
                "type": "span",
              },
            ],
            "type": "button",
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="suggestions--li--button"
              onClick={[Function]}
>
              <span>
                            dames 
                            <span
                                          className="suggestions--li--button--span"
                            >
                                          trui
                            </span>
                            en
              </span>
               
              <span
                            className="suggestions--li--button--span"
              >
                            (
                            1501
                            )
              </span>
</button>,
            "className": "suggestions--li",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span>
                  dames 
                  <span
                                    className="suggestions--li--button--span"
                  >
                                    trui
                  </span>
                  en
</span>,
                " ",
                <span
                  className="suggestions--li--button--span"
>
                  (
                  1501
                  )
</span>,
              ],
              "className": "suggestions--li--button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "dames ",
                    <span
                      className="suggestions--li--button--span"
>
                      trui
</span>,
                    "en",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "dames ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "trui",
                      "className": "suggestions--li--button--span",
                    },
                    "ref": null,
                    "rendered": "trui",
                    "type": "span",
                  },
                  "en",
                ],
                "type": "span",
              },
              " ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "(",
                    1501,
                    ")",
                  ],
                  "className": "suggestions--li--button--span",
                },
                "ref": null,
                "rendered": Array [
                  "(",
                  1501,
                  ")",
                ],
                "type": "span",
              },
            ],
            "type": "button",
          },
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
